torchtext/experimental/datasets/sequence_tagging.py

```diff
@@ -0,0 +1,155 @@
+import torch
+
+from torchtext.experimental.datasets import raw
+from torchtext.vocab import build_vocab_from_iterator
+from torchtext.experimental.functional import (
+    vocab_func,
+    totensor,
+    sequential_transforms,
+)
+
+
+def _build_vocab(data):
+    total_columns = len(data[0])
+    data_list = [[] for _ in range(total_columns)]
+    vocabs = []
+
+    for line in data:
+        for idx, col in enumerate(line):
+            data_list[idx].append(col)
+
+    for it in data_list:
+        vocabs.append(build_vocab_from_iterator(it))
+
+    return vocabs
+
+
+def _setup_datasets(dataset_name,
+                    root=".data",
+                    vocabs=None,
+                    data_select=("train", "valid", "test")):
+    train, val, test = DATASETS[dataset_name](root=root)
+    raw_data = {
+        "train": [line for line in train] if train else None,
+        "valid": [line for line in val] if val else None,
+        "test": [line for line in test] if test else None
+    }
+
+    if vocabs is None:
+        if "train" not in data_select:
+            raise TypeError("Must pass a vocab if train is not selected.")
+        vocabs = _build_vocab(raw_data["train"])
+    else:
+        if not isinstance(vocabs, list):
+            raise TypeError("vocabs must be an instance of list")
+
+        # Find data that's not None
+        notnone_data = None
+        for key in raw_data.keys():
+            if raw_data[key] is not None:
+                notnone_data = raw_data[key]
+                break
+        if len(vocabs) != len(notnone_data[0]):
+            raise ValueError(
+                "Number of vocabs must match the number of columns "
+                "in the data")
+
+    transformers = [
+        sequential_transforms(vocab_func(vocabs[idx]),
+                              totensor(dtype=torch.long))
+        for idx in range(len(vocabs))
+    ]
+
+    datasets = []
+    for item in data_select:
+        if raw_data[item] is not None:
+            datasets.append(
+                SequenceTaggingDataset(raw_data[item], vocabs, transformers))
+
+    return datasets
+
+
+class SequenceTaggingDataset(torch.utils.data.Dataset):
+    """Defines an abstraction for raw text sequence tagging iterable datasets.
+    Currently, we only support the following datasets:
+        - UDPOS
+        - CoNLL2000Chunking
+    """
+    def __init__(self, data, vocab, transforms):
+        """Initiate sequence tagging dataset.
+        Arguments:
+            data: a list of word and its respective tags. Example:
+                [[word, POS, dep_parsing label, ...]]
+            vocab: Vocabulary object used for dataset.
+            transforms: a list of string transforms for words and tags.
+        """
+
+        super(SequenceTaggingDataset, self).__init__()
+        self.data = data
+        self.vocab = vocab
+        self.transforms = transforms
+
+    def __getitem__(self, i):
```

Comment:
Should we check that the data[i] and transforms have same number of columns here?

Reply:
done